HerPublicWebsite.BusinessLogic/ExternalServices/OsPlaces/OsPlacesApi.cs

```diff
@@ -1,33 +1,52 @@
+using HerPublicWebsite.BusinessLogic.Extensions;
+using HerPublicWebsite.BusinessLogic.ExternalServices.Common;
 using HerPublicWebsite.BusinessLogic.Models;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
 
 namespace HerPublicWebsite.BusinessLogic.ExternalServices.OsPlaces;
 
 public class OsPlacesApi : IOsPlacesApi
 {
-    public async Task<List<Address>> GetAddressesAsync(string postcode, string buildingNumber)
+    private readonly OsPlacesConfiguration config;
+    private readonly ILogger<OsPlacesApi> logger;
+
+    public OsPlacesApi(IOptions<OsPlacesConfiguration> options, ILogger<OsPlacesApi> logger)
+    {
+        this.config = options.Value;
+        this.logger = logger;
+    }
+    
+    public async Task<List<Address>> GetAddressesAsync(string postcode, string buildingNameOrNumber)
     {
-        //TODO BEISHER-248: Actually implement API here
-        return new List<Address> {
-            new () {
-                DisplayAddress = "82 My House, My Area, A Special Location, LB6 420",
-                AddressLine1 = "82 My House",
-                AddressLine2 = "My Area",
-                Postcode = "LB6 420",
-                County = "A County",
-                Town = "Camdoon",
-                Uprn = "023232323",
-                LocalCustodianCode = "2323",
-            },
-            new ()
+        if (!postcode.IsValidUkPostcodeFormat())
+        {
+            return new List<Address>();
+        }
+        
+        var parameters = new RequestParameters
+        {
+            BaseAddress = config.BaseUrl,
+            Path = $"/search/places/v1/postcode?postcode={postcode}",
+            Headers = new Dictionary<string, string> { { "Key", config.Key } }
+        };
+
+        try {
+            var response = await HttpRequestHelper.SendGetRequestAsync<OsPlacesPostcodeResponseDto>(parameters);
+
+            var filteredResults = response.Results.Where(r => r.Dpa.BuildingNumber == buildingNameOrNumber || r.Dpa.BuildingName?.ToLower() == buildingNameOrNumber.ToLower()).ToList();
```

Comment:
We need to check for empty results here first - this will throw if `Results` is `null`

Reply:
Hmm, not sure if that's possible or not, but no harm in covering it.

Reply:
I encountered it when putting in a postcode with no results :)